src/cogent3/app/comp_new.py

```diff
@@ -0,0 +1,744 @@
+U = TypeVar("U")
+
+P = ParamSpec("P")
+Q = ParamSpec("Q")
+
+R = TypeVar("R")
+S = TypeVar("S")
+
+
+class AppType(Enum):
+    LOADER = "loader"
+    WRITER = "writer"
+    GENERIC = "generic"
+    NON_COMPOSABLE = "non_composable"
+
+
+# Aliases to use Enum easily
+LOADER = AppType.LOADER
+WRITER = AppType.WRITER
+GENERIC = AppType.GENERIC
+NON_COMPOSABLE = AppType.NON_COMPOSABLE
+
+
+class _AppBaseClass(Generic[T, P, R]):
+    fn: Callable[Concatenate[T, P], R]
+    app_type: ClassVar[AppType]
+    original_name: ClassVar[str]
+    is_composed: ClassVar[bool] = False
+    skip_not_completed: bool = True
+
+    def __init__(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        self.args = args
+        self.kwargs = kwargs
+        self.type_check: bool = True
+        self.raise_errors: bool = False
+        self.write_data_store = None
+        if self.app_type is WRITER:
+            for arg in self.args:
+                if isinstance(arg, DataStoreABC):
+                    self.write_data_store = arg
+                    break
+            if self.write_data_store is None:
+                for kwarg in self.kwargs.values():
+                    if isinstance(kwarg, DataStoreABC):
+                        self.write_data_store = kwarg
+                        break
+            if self.write_data_store is None:
+                msg = "A writer app must initialise a writable datastore."
+                raise ValueError(msg)
+
+    def set_type_check(self, type_check: bool) -> None:
+        self.type_check = type_check
+
+    def set_raise_errors(self, raise_errors: bool) -> None:
+        self.raise_errors = raise_errors
+
+    def __call__(self, value: T | NotCompleted) -> R | NotCompleted:
+        print("Calling")
+        print("The app is", self.fn.__name__)
+        if isinstance(value, NotCompleted) and self.skip_not_completed:
+            return value
+        cls = type(self)
+
+        if self.type_check:
+            type_hints = get_type_hints(cls.fn)
+            if not isinstance(value, NotCompleted):
+                check_type(value, type_hints[next(iter(type_hints))])
+        else:
+            type_hints = {}
+        print("Doing function")
+
+        try:
+            result = cls.fn(cast("T", value), *self.args, **self.kwargs)
+        except Exception as e:
+            if isinstance(e, _AppInstantiationError):
+                raise e.original.with_traceback(e.tb) from None
+            if self.raise_errors:
+                raise
+            result = NotCompleted("ERROR", self, traceback.format_exc(), source=value)
+
+        print("FINISHED CALL, GOT", type(result))
+
+        if self.type_check and not isinstance(result, NotCompleted):
+            print("TYPE IS", type(result), "EXPECTED", type_hints["return"])
+            check_type(result, type_hints["return"])
+        return result
+
+    def main(self, value: T) -> R:
+        raise_errors = self.raise_errors
+        self.raise_errors = True
```

Comment:
Looks like a debugging condition? If so, probably best to delete this line and make the default app behaviour `raise_errors = True`.